logstash-core/src/main/java/org/logstash/execution/WorkerLoop.java

```diff
@@ -17,42 +13,43 @@
 
     private final Dataset execution;
 
-    private final BlockingQueue<IRubyObject> signalQueue;
-
     private final QueueReadClient readClient;
 
+    private final AtomicBoolean flushRequested;
+
     private final AtomicBoolean flushing;
 
+    private final AtomicBoolean shutdownRequested;
+
     private final LongAdder consumedCounter;
 
     private final LongAdder filteredCounter;
 
     private final boolean drainQueue;
 
-    public WorkerLoop(final CompiledPipeline pipeline, final BlockingQueue<IRubyObject> signalQueue,
-        final QueueReadClient readClient, final LongAdder filteredCounter,
-        final LongAdder consumedCounter, final AtomicBoolean flushing, final boolean drainQueue) {
+    public WorkerLoop(final CompiledPipeline pipeline, final QueueReadClient readClient,
+        final LongAdder filteredCounter, final LongAdder consumedCounter,
+        final AtomicBoolean flushRequested, final AtomicBoolean flushing,
+        final AtomicBoolean shutdownRequested, final boolean drainQueue) {
         this.consumedCounter = consumedCounter;
         this.filteredCounter = filteredCounter;
         this.execution = pipeline.buildExecution();
-        this.signalQueue = signalQueue;
         this.drainQueue = drainQueue;
         this.readClient = readClient;
+        this.flushRequested = flushRequested;
         this.flushing = flushing;
+        this.shutdownRequested = shutdownRequested;
     }
 
     @Override
     public void run() {
         try {
-            boolean shutdownRequested = false;
-            final ThreadContext context = RubyUtil.RUBY.getCurrentContext();
+            boolean isShutdown;
             do {
-                final IRubyObject signal = signalQueue.poll();
-                shutdownRequested = shutdownRequested
-                    || signal != null && signal.callMethod(context, "shutdown?").isTrue();
+                isShutdown = shutdownRequested.get();
```

Comment:
@danhermann I think you we should keep with the old behaviour and simply do:

`isShutdown = isShutdown || shutdownRequested.get();`

That's what I was aiming at to prevent being the possibility of "reverting" a shutdown so to speak. Sorry about the confusion :)